File: src/main/java/br/com/augustoccesar/querybuilder/query/Join.java
package br.com.augustoccesar.querybuilder.query;

import java.util.Arrays;

/**
 * Created by augustoccesar on 6/2/16.
 */
public class Join {
    public enum Type {
        LEFT_JOIN("LEFT JOIN"),
        RIGHT_JOIN("RIGHT JOIN"),
        INNER_JOIN("INNER JOIN");

        private String value;

        Type(String value) {
            this.value = value;
        }

        public String getValue() {
            return value;
        }
    }

    public Type type;
    public Table table;
    public String leftJoinOn;
    public String rightJoinOn;

    public static final Type INNER = Type.INNER_JOIN;
    public static final Type LEFT = Type.LEFT_JOIN;
    public static final Type RIGHT = Type.RIGHT_JOIN;

    // Constructors

    public Join(Type type) {
        this.type = type;
    }

    public Join(Type type, String tableOrTableMarkdown, String joinOn) {
        this.type = type;
        this.table = new Table(tableOrTableMarkdown);
        this.splitJoinOn(joinOn);
    }

    public Join(Type type, String tableOrTableMarkdown, String leftJoinOn, String rightJoinOn) {
        this.type = type;
        this.table = new Table(tableOrTableMarkdown);
        this.leftJoinOn = leftJoinOn;
        this.rightJoinOn = rightJoinOn;
    }

    // Builder

    public Join table(String tableOrTableMarkdown) {
        this.table = new Table(tableOrTableMarkdown);
        return this;
    }

    public Join on(String joinOn){
        this.splitJoinOn(joinOn);
        return this;
    }

    public Join on(String leftJoinOn, String rightJoinOn) {
        this.leftJoinOn = leftJoinOn;
        this.rightJoinOn = rightJoinOn;
        return this;
    }

    // Helpers

    private void splitJoinOn(String joinOn){
        String[] result = joinOn.split("=");
        Arrays.asList(result).forEach((s) -> s = s.replace(" ", ""));

        this.leftJoinOn = result[0];
        this.rightJoinOn = result[1];
    }
}

File: src/main/java/br/com/augustoccesar/querybuilder/query/Join.kt
package br.com.augustoccesar.querybuilder.query

import br.com.augustoccesar.querybuilder.builders.Buildable
import br.com.augustoccesar.querybuilder.constants.CommonStrings
import java.util.*

/**
 * Created by augustoccesar on 6/2/16.
 */
class Join : Buildable {
    enum class Type constructor(val value: String) {
        LEFT_JOIN("LEFT JOIN"),
        RIGHT_JOIN("RIGHT JOIN"),
        INNER_JOIN("INNER JOIN")
    }

    private var type: Type? = null
    private var table: Table? = null
    private var leftJoinOn: Column? = null
    private var rightJoinOn: Column? = null

    constructor(type: Type) {
        this.type = type
    }

    constructor(type: Type, tableOrTableMarkdown: String, joinOn: String) {
        this.type = type
        this.table = Table.fromMarkdown(tableOrTableMarkdown)
        this.splitJoinOn(joinOn)
    }

    constructor(type: Type, tableOrTableMarkdown: String, leftJoinOn: String, rightJoinOn: String) {
        this.type = type
        this.table = Table.fromMarkdown(tableOrTableMarkdown)
        this.leftJoinOn = Column.fromMarkdown(leftJoinOn)
        this.rightJoinOn = Column.fromMarkdown(rightJoinOn)
    }

    fun table(tableOrTableMarkdown: String): Join {
        this.table = Table.fromMarkdown(tableOrTableMarkdown)
        return this
    }

    fun on(joinOn: String): Join {
        this.splitJoinOn(joinOn)
        return this
    }

    fun on(leftJoinOn: String, rightJoinOn: String): Join {
        this.leftJoinOn = Column.fromMarkdown(leftJoinOn)
        this.rightJoinOn = Column.fromMarkdown(rightJoinOn)
        return this
    }

    private fun splitJoinOn(joinOn: String) {
        val result = joinOn.split("=".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        Arrays.asList(*result).forEach { s -> s.replace(" ", "") }

        this.leftJoinOn = Column.fromMarkdown(result[0].replace(" ", ""))
        this.rightJoinOn = Column.fromMarkdown(result[1].replace(" ", ""))
    }

    override fun build(): String {
        val stringBuilder = StringBuilder()

        stringBuilder.append(" ")
                .append(this.type!!.value).append(" ")
                .append(this.table!!.build()).append(CommonStrings.ON)
                .append(this.leftJoinOn!!.sqlColumnRepresentation()).append(Comparison.EQUALS.value).append(this.rightJoinOn!!.sqlColumnRepresentation())
                .append(" ")

        return stringBuilder.toString().replace("\\s+".toRegex(), " ")
    }

    companion object {
        @JvmField
        val INNER = Type.INNER_JOIN
        @JvmField
        val LEFT = Type.LEFT_JOIN
        @JvmField
        val RIGHT = Type.RIGHT_JOIN
    }
}
